### ISSUE TITLE
RFC: offset/weight handling (mixlm compatibility problem)

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#191]
In the current version (1.1-5) we have managed to break compatibility with the `mixlm` package.

**tl;dr** I've fixed the problem, sort of, but this exposes a bit of the code that is fragile; there might be a much less ugly way to solve the original problem ... ?

```
library(mixlm)
packageVersion("mixlm") ##  1.0.0
packageVersion("lme4")  ##  1.1.5
set.seed(101)
dataset   <- data.frame(y=rnorm(8),
       x=factor(c(rep(1,4),rep(0,4))),
       z=factor(rep(c(1,0),4)))
GLM       <- glm(y  ~ x+r(z), family=gaussian(identity), data=dataset)
## Error in get(i, parent.frame()) : 
##   argument "weights" is missing, with no default
## In addition: Warning message:
## In lme4::glmer(rw$reml.formula, data, family = family, REML = REML,  :
##   calling glmer() with family=gaussian (identity link) as a shortcut to lmer() is deprecated; please call lmer() directly
```

Proximally, the problem occurs (actually "occurred -- now fixed, see below) in `lFormula` (in `R/modular.R`), at l. 295:

```
## model.frame.default looks for these objects in the environment
## of the *formula* (see 'extras', which is anything passed in ...),
## so they have to be put there ...
for (i in c("weights", "offset")) {
     assign(i,get(i,parent.frame()),environment(fr.form))
}
```

No-one has reported problems with this in the past, but any time I muck around with getting things out of parent frames I accept that it can be fragile.  The particular problem here is that normally when `weights` is missing we end(ed) up doing something silly but harmless -- the `weights()` function from `stats` was copied.  In this case we are in trouble because the parent frame is a function with a `weights` argument that might be missing (no default/not specified by user), so trying to evaluate it throws an error.

This code is now fixed as follows:

```
for (i in c("weights", "offset")) {
    if (!eval(bquote(missing(x=.(i)))))
        assign(i,get(i,parent.frame()),environment(fr.form))
}
```

so that it doesn't try to assign the variable if it hasn't been passed to `lFormula`.

However, I wonder if there is an entirely better way to do this.  Indeed, I don't remember why it's written this way in the first place, but there clearly must have been a reason why we had to use `get(i,parent.frame())` (rather than just `get(i)`) ...
